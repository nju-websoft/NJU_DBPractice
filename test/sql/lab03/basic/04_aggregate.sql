open database db2025;

-- 1. Aggregate functions: COUNT, SUM, AVG, MIN, MAX.
select count(*) from stock;
select count(s_i_id) from stock;
select count(s_i_id) from stock where s_w_id = 751;
select count(s_i_id) from stock where s_w_id >=10 and s_quantity < 50;
select count(s_i_id) from stock where s_w_id >=500 and s_quantity < 25;
select sum(s_quantity) from stock;
select avg(s_quantity) from stock;
select min(s_quantity) from stock;
select max(s_quantity) from stock;
select count(i_im_id) from stock outer join item where stock.s_i_id = item.i_im_id order by s_i_id;
select sum(i_im_id) from stock outer join item where stock.s_i_id = item.i_im_id order by s_i_id;
select avg(i_im_id) from stock outer join item where stock.s_i_id = item.i_im_id order by s_i_id;
select min(i_im_id) from stock outer join item where stock.s_i_id = item.i_im_id order by s_i_id;
select max(i_im_id) from stock outer join item where stock.s_i_id = item.i_im_id order by s_i_id;

-- 2. Grouping results with GROUP BY.
select s_quantity, count(*) from stock group by s_quantity order by s_quantity;
select s_quantity, sum(s_remote_cnt) from stock group by s_quantity order by s_quantity;
select s_quantity, avg(s_remote_cnt) from stock group by s_quantity order by s_quantity;

-- 3. test having clause with aggregate functions.
select s_quantity, count(*) from stock group by s_quantity having count(*) > 10 order by s_quantity;
select s_quantity, i_im_id, count(*) from (select * from stock join item where s_i_id=i_im_id) group by s_quantity, i_im_id having count(*)>=3 order by s_quantity, i_im_id;
select s_quantity, sum(s_remote_cnt) from stock group by s_quantity having sum(s_remote_cnt) > 100 order by s_quantity;
select s_quantity, avg(s_remote_cnt), sum(s_order_cnt) from stock group by s_quantity having sum(s_order_cnt) > 200 order by s_quantity;
select s_quantity, avg(s_remote_cnt) as AVG_remote, avg(s_order_cnt), max(s_order_cnt), min(s_ytd), count(*) from stock where s_ytd > 5 group by s_quantity having sum(s_order_cnt) > 200 order by s_quantity;
select s_quantity, count(i_im_id), max(i_price), min(i_price), sum(i_price), count(i_price), count(*) from (select * from stock outer join item where s_i_id=i_im_id) where s_ytd > 3 group by s_quantity having s_quantity > 3 and avg(s_ytd) >= 6.93 order by s_quantity;