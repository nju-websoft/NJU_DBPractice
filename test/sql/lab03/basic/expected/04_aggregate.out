
+--------------+
| COUNT(*)     | 
+--------------+
| 100          | 
+--------------+
Total tuple(s): 1

+---------------+
| COUNT(s_i_id) | 
|               | 
+---------------+
| 100           | 
+---------------+
Total tuple(s): 1

+---------------+
| COUNT(s_i_id) | 
|               | 
+---------------+
| 1             | 
+---------------+
Total tuple(s): 1

+---------------+
| COUNT(s_i_id) | 
|               | 
+---------------+
| 100           | 
+---------------+
Total tuple(s): 1

+---------------+
| COUNT(s_i_id) | 
|               | 
+---------------+
| 55            | 
+---------------+
Total tuple(s): 1

+-----------------+
| SUM(s_quantity) | 
|                 | 
+-----------------+
| 528             | 
+-----------------+
Total tuple(s): 1

+-----------------+
| AVG(s_quantity) | 
|                 | 
+-----------------+
| 5               | 
+-----------------+
Total tuple(s): 1

+-----------------+
| MIN(s_quantity) | 
|                 | 
+-----------------+
| 0               | 
+-----------------+
Total tuple(s): 1

+-----------------+
| MAX(s_quantity) | 
|                 | 
+-----------------+
| 10              | 
+-----------------+
Total tuple(s): 1

+----------------+
| COUNT(i_im_id) | 
|                | 
+----------------+
| 100            | 
+----------------+
Total tuple(s): 1

+--------------+
| SUM(i_im_id) | 
|              | 
+--------------+
| 2408         | 
+--------------+
Total tuple(s): 1

+--------------+
| AVG(i_im_id) | 
|              | 
+--------------+
| 24           | 
+--------------+
Total tuple(s): 1

+--------------+
| MIN(i_im_id) | 
|              | 
+--------------+
| 0            | 
+--------------+
Total tuple(s): 1

+--------------+
| MAX(i_im_id) | 
|              | 
+--------------+
| 49           | 
+--------------+
Total tuple(s): 1

+--------------+--------------+
| s_quantity   | COUNT(*)     | 
+--------------+--------------+
| 0            | 6            | 
+--------------+--------------+
| 1            | 8            | 
+--------------+--------------+
| 2            | 10           | 
+--------------+--------------+
| 3            | 12           | 
+--------------+--------------+
| 4            | 7            | 
+--------------+--------------+
| 5            | 10           | 
+--------------+--------------+
| 6            | 10           | 
+--------------+--------------+
| 7            | 6            | 
+--------------+--------------+
| 8            | 9            | 
+--------------+--------------+
| 9            | 8            | 
+--------------+--------------+
| 10           | 14           | 
+--------------+--------------+
Total tuple(s): 11

+--------------+-------------------+
| s_quantity   | SUM(s_remote_cnt) | 
|              |                   | 
+--------------+-------------------+
| 0            | 394               | 
+--------------+-------------------+
| 1            | 512               | 
+--------------+-------------------+
| 2            | 653               | 
+--------------+-------------------+
| 3            | 683               | 
+--------------+-------------------+
| 4            | 379               | 
+--------------+-------------------+
| 5            | 528               | 
+--------------+-------------------+
| 6            | 537               | 
+--------------+-------------------+
| 7            | 270               | 
+--------------+-------------------+
| 8            | 459               | 
+--------------+-------------------+
| 9            | 438               | 
+--------------+-------------------+
| 10           | 692               | 
+--------------+-------------------+
Total tuple(s): 11

+--------------+-------------------+
| s_quantity   | AVG(s_remote_cnt) | 
|              |                   | 
+--------------+-------------------+
| 0            | 65                | 
+--------------+-------------------+
| 1            | 64                | 
+--------------+-------------------+
| 2            | 65                | 
+--------------+-------------------+
| 3            | 56                | 
+--------------+-------------------+
| 4            | 54                | 
+--------------+-------------------+
| 5            | 52                | 
+--------------+-------------------+
| 6            | 53                | 
+--------------+-------------------+
| 7            | 45                | 
+--------------+-------------------+
| 8            | 51                | 
+--------------+-------------------+
| 9            | 54                | 
+--------------+-------------------+
| 10           | 49                | 
+--------------+-------------------+
Total tuple(s): 11

+--------------+--------------+
| s_quantity   | COUNT(*)     | 
+--------------+--------------+
| 3            | 12           | 
+--------------+--------------+
| 10           | 14           | 
+--------------+--------------+
Total tuple(s): 2

+--------------+--------------+--------------+
| s_quantity   | i_im_id      | COUNT(*)     | 
+--------------+--------------+--------------+
| 2            | 16           | 3            | 
+--------------+--------------+--------------+
| 3            | 2            | 5            | 
+--------------+--------------+--------------+
| 3            | 12           | 4            | 
+--------------+--------------+--------------+
| 3            | 47           | 3            | 
+--------------+--------------+--------------+
| 3            | 48           | 4            | 
+--------------+--------------+--------------+
| 3            | 49           | 3            | 
+--------------+--------------+--------------+
| 5            | 11           | 4            | 
+--------------+--------------+--------------+
| 6            | 4            | 4            | 
+--------------+--------------+--------------+
| 6            | 40           | 3            | 
+--------------+--------------+--------------+
| 7            | 39           | 3            | 
+--------------+--------------+--------------+
| 7            | 44           | 3            | 
+--------------+--------------+--------------+
| 9            | 3            | 3            | 
+--------------+--------------+--------------+
| 9            | 22           | 3            | 
+--------------+--------------+--------------+
| 10           | 19           | 5            | 
+--------------+--------------+--------------+
| 10           | 46           | 3            | 
+--------------+--------------+--------------+
Total tuple(s): 15

+--------------+-------------------+
| s_quantity   | SUM(s_remote_cnt) | 
|              |                   | 
+--------------+-------------------+
| 0            | 394               | 
+--------------+-------------------+
| 1            | 512               | 
+--------------+-------------------+
| 2            | 653               | 
+--------------+-------------------+
| 3            | 683               | 
+--------------+-------------------+
| 4            | 379               | 
+--------------+-------------------+
| 5            | 528               | 
+--------------+-------------------+
| 6            | 537               | 
+--------------+-------------------+
| 7            | 270               | 
+--------------+-------------------+
| 8            | 459               | 
+--------------+-------------------+
| 9            | 438               | 
+--------------+-------------------+
| 10           | 692               | 
+--------------+-------------------+
Total tuple(s): 11

+--------------+-------------------+------------------+
| s_quantity   | AVG(s_remote_cnt) | SUM(s_order_cnt) | 
|              |                   |                  | 
+--------------+-------------------+------------------+
| 1            | 64                | 365              | 
+--------------+-------------------+------------------+
| 2            | 65                | 506              | 
+--------------+-------------------+------------------+
| 3            | 56                | 677              | 
+--------------+-------------------+------------------+
| 4            | 54                | 343              | 
+--------------+-------------------+------------------+
| 5            | 52                | 387              | 
+--------------+-------------------+------------------+
| 6            | 53                | 719              | 
+--------------+-------------------+------------------+
| 7            | 45                | 398              | 
+--------------+-------------------+------------------+
| 8            | 51                | 508              | 
+--------------+-------------------+------------------+
| 9            | 54                | 424              | 
+--------------+-------------------+------------------+
| 10           | 49                | 775              | 
+--------------+-------------------+------------------+
Total tuple(s): 10

+--------------+--------------+------------------+------------------+--------------+--------------+
| s_quantity   | AVG_remote   | AVG(s_order_cnt) | MAX(s_order_cnt) | MIN(s_ytd)   | COUNT(*)     | 
|              |              |                  |                  |              |              | 
+--------------+--------------+------------------+------------------+--------------+--------------+
| 2            | 61           | 56               | 88               | 6.229845     | 7            | 
+--------------+--------------+------------------+------------------+--------------+--------------+
| 3            | 41           | 54               | 69               | 5.649854     | 5            | 
+--------------+--------------+------------------+------------------+--------------+--------------+
| 4            | 42           | 59               | 97               | 6.835971     | 5            | 
+--------------+--------------+------------------+------------------+--------------+--------------+
| 6            | 34           | 60               | 95               | 5.320866     | 5            | 
+--------------+--------------+------------------+------------------+--------------+--------------+
| 8            | 58           | 57               | 88               | 5.208548     | 4            | 
+--------------+--------------+------------------+------------------+--------------+--------------+
| 9            | 52           | 57               | 76               | 5.693364     | 4            | 
+--------------+--------------+------------------+------------------+--------------+--------------+
| 10           | 50           | 68               | 89               | 5.337111     | 5            | 
+--------------+--------------+------------------+------------------+--------------+--------------+
Total tuple(s): 7

+--------------+----------------+--------------+--------------+--------------+----------------+--------------+
| s_quantity   | COUNT(i_im_id) | MAX(i_price) | MIN(i_price) | SUM(i_price) | COUNT(i_price) | COUNT(*)     | 
|              |                |              |              |              |                |              | 
+--------------+----------------+--------------+--------------+--------------+----------------+--------------+
| 4            | 3              | 9.513620     | 0.286370     | 14.372331    | 3              | 6            | 
+--------------+----------------+--------------+--------------+--------------+----------------+--------------+
| 6            | 6              | 3.643920     | 0.122230     | 8.692511     | 6              | 10           | 
+--------------+----------------+--------------+--------------+--------------+----------------+--------------+
| 7            | 3              | 6.122020     | 1.543930     | 12.040689    | 3              | 4            | 
+--------------+----------------+--------------+--------------+--------------+----------------+--------------+
| 9            | 4              | 7.210390     | 1.457610     | 14.694750    | 4              | 6            | 
+--------------+----------------+--------------+--------------+--------------+----------------+--------------+
Total tuple(s): 4
